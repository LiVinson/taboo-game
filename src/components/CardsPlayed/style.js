import styled from "styled-components"

export const NoCardMessage = styled.p`
	font-size:2rem;
	color: ${props => props.theme.color.grayDark2};
`