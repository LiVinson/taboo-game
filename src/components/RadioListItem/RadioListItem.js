import React from 'react'
import PropTypes from 'prop-types'
import { StyledRadioListItem, RadioInput, RadioLabel } from './style'

//Takes in taboo word, it's index, status, and a cb function that is called when user selects a different word in the list
//Returns a list item with a hidden radio button and label.
const RadioListItem = ({ word, index, status, checked, onChange }) => {
	return (
		<StyledRadioListItem>
			<RadioInput
				checked={checked}
				type="radio"
				name={status}
				value={index}
				id={word}
				onChange={(e) => {
					onChange(e.target.value, status)
				}}
			/>
			<RadioLabel htmlFor={word}>{word}</RadioLabel>
		</StyledRadioListItem>
	)
}

RadioListItem.propTypes = {
	word: PropTypes.string.isRequired,
	index: PropTypes.number.isRequired,
	status: PropTypes.string.isRequired,
	onChange: PropTypes.func.isRequired,
}
export default RadioListItem
