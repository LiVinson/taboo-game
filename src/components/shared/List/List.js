import React from "react"
import {StyledList} from "./style"

const List = ({children}) => (
    <StyledList>
        {children}
    </StyledList>
)

export default List