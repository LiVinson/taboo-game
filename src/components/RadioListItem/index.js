import RadioListItem from "./RadioListItem"

export default RadioListItem