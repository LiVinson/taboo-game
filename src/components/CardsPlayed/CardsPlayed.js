import React from 'react'
import PropTypes from 'prop-types'
import { ButtonTabooCard } from 'components/shared/TabooCard'
import RadioListItem from 'components/RadioListItem'
import List from 'components/shared/List'
import { NoCardMessage } from './style'

//Returns a Taboo card with buttons to change card status to other two options or a card stating there are no cards in this status
//status is correct, skipped or discarded with cardList an array of all cards from current round in this status.
//Selected is string of currently selected cards index. Used to determine which radio button to check
const CardsPlayed = ({ status, cardList, selected, handleChange, buttonInfo, error }) => {
	return (
		<ButtonTabooCard tabooWord={status} buttons={buttonInfo} error={error}>
			<List>
				{cardList.length > 0 ? (
					cardList.map((card) => {
						return (
							<RadioListItem
								key={card.index} //index in deck in firestore which is unchanging
								word={card.word}
								index={card.index}
								onChange={handleChange}
								status={status}
								checked={String(card.index) === selected}
							/>
						)
					})
				) : (
					<NoCardMessage>No {status} cards this round</NoCardMessage>
				)}
			</List>
		</ButtonTabooCard>
	)
}

CardsPlayed.propTypes = {
	status: PropTypes.string.isRequired,
	cardList: PropTypes.array.isRequired,
	selected: PropTypes.string.isRequired,
	handleChange: PropTypes.func.isRequired,
	buttonInfo: PropTypes.array.isRequired,
	error: PropTypes.string,
}

export default CardsPlayed
