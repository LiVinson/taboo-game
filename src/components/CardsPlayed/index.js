import CardsPlayed from "./CardsPlayed"

export default CardsPlayed