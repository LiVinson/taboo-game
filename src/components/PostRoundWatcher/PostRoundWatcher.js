import React from 'react'
import PropTypes from 'prop-types'
import InstructionsText from 'components/shared/InstructionsText'
import CardsPlayed from 'components/CardsPlayed'
import { LargeButton } from 'components/shared/Button'
import Pending from 'components/shared/Pending'
import { ErrorMessage } from 'components/shared/FeedbackMessage'
import KeyWord from 'components/shared/KeyWord'

//Used for each of 3 status types to create buttons to change card status to the other 2 statuses
export const generateCardsPlayedButtonInfo = (statusArray, status, statusSelected, cb, isPending) => {
	const buttonInfo = statusArray
		.filter((cardType) => cardType !== status)
		.map((cardType) => {
			const button = { className: 'button' }
			button.text = cardType
			button.disabled = statusSelected.length < 1 || isPending ? true : false
			button.onClick = () => cb(status, cardType)
			return button
		})

	return buttonInfo
}

const PostRoundWatcher = ({
	cardStatuses,
	cardsPlayed,
	selection,
	handleCardSelection,
	selectedCards,
	updateSelectedCard,
	confirmRoundEnd,
	isPending,
	error,
}) => {
	
	//Loops over each card status, and returns a Taboo card with 2 buttons.
	//Filters cards played based on the status. Information needed to generate buttons,
	//select a card in the list, and change button status passed in
	// Then separates cards played in round based on current card status
	return (
		<React.Fragment>
			<InstructionsText align="center">
				Review the cards from this round below. Select the word and click the button to change it's status.
				Select <KeyWord>Confirm!</KeyWord> when done to calculate scores.
			</InstructionsText>
			{cardStatuses.map((status, index) => (
				<CardsPlayed
					// Only pass down error message if it is for this Card status. Indicated by error message include status type
					error={
						error.cards &&
						cardsPlayed.some(
							(card) => card.status === status && error.cards.includes(card.index.toString())
						)
							? error.cards
							: null
					}
					key={status}
					status={cardStatuses[index]}
					selected={selectedCards[index]}
					cardList={cardsPlayed.filter((card) => card.status === status)}
					handleChange={handleCardSelection}
					buttonInfo={generateCardsPlayedButtonInfo(
						cardStatuses,
						status,
						selectedCards[index], //selectedCards array order is same as status array order, so indexes correspond
						updateSelectedCard,
						isPending.cards //true when in progress of updating status
					)}
				/>
			))}
			<LargeButton text="Confirm!" disabled={isPending.round} onClick={confirmRoundEnd} />
			{isPending.round ? <Pending speed={300} message={'Updating scores'} /> : null}
			{error.round && <ErrorMessage error={error.round} />}
		</React.Fragment>
	)
}

PostRoundWatcher.propTypes = {
	cardStatuses: PropTypes.array.isRequired,
	cardsPlayed: PropTypes.array.isRequired,
	handleCardSelection: PropTypes.func.isRequired,
	selectedCards: PropTypes.array.isRequired,
	updateSelectedCard: PropTypes.func.isRequired,
	confirmRoundEnd: PropTypes.func.isRequired,
	isPending: PropTypes.object.isRequired,
	error: PropTypes.object.isRequired,
}

export default PostRoundWatcher
