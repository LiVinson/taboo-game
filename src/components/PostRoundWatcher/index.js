import PostRoundWatcher from "./PostRoundWatcher"

export default PostRoundWatcher