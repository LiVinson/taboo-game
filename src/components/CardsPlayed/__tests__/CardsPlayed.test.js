import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { ThemeProvider } from 'styled-components'
import theme from '../../../global-design/theme'
import CardsPlayed from '../CardsPlayed'
import { ButtonTabooCard } from 'components/shared/TabooCard'
import RadioListItem from 'components/RadioListItem'
import List from 'components/shared/List'
import { cardsPlayed } from '__fixtures__/deck'
import { generateCardsPlayedButtonInfo } from 'components/PostRoundWatcher/PostRoundWatcher'

describe('CardsPlayed', () => {
	const cardStatuses = ['correct', 'skipped', 'discard']
	const buttonInfo = generateCardsPlayedButtonInfo(cardStatuses, 'correct', '', jest.fn(), false)
	const defaultProps = {
		status: 'correct',
		cardList: cardsPlayed.filter(card => card.status === "correct"),
		handleChange: jest.fn(),
		buttonInfo,
		error: null,
		selected: ""
	}

	test('Snapshot', () => {
		const props = {
			...defaultProps,
		}

		const wrapper = renderer
			.create(
				<ThemeProvider theme={theme}>
					<CardsPlayed {...props} />
				</ThemeProvider>
			)
			.toJSON()
		expect(wrapper).toMatchSnapshot()
	})

	test('Renders a ButtonTabooCard', () => {
		const props = {
			...defaultProps,
		}

		const wrapper = shallow(<CardsPlayed {...props} />)
		expect(wrapper.find(ButtonTabooCard)).toHaveLength(1)
	})

	test('Renders a RadioListItem for each card in the status category', () => {
		const props = {
			...defaultProps,
		}

        const wrapper = shallow(<CardsPlayed {...props} />)
        expect(wrapper.find(ButtonTabooCard).dive().find(List).dive().find(RadioListItem)).toHaveLength(2)
	})

	test('Renders NoCardMessage when no cards in the specified status', () => {
		const props = {
            ...defaultProps,
            cardList: defaultProps.cardList.filter(card => card.status !== "correct")
		}
        const wrapper = shallow(<CardsPlayed {...props} />)
        expect(wrapper.find(ButtonTabooCard).dive().find(List).shallow().text()).toBe("No correct cards this round")

	})
})
