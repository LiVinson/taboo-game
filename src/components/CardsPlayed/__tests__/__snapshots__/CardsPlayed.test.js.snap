// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardsPlayed Snapshot 1`] = `
<div
  className="sc-AxjAm hDlFTS"
>
  <div
    className="sc-AxiKw dGnLMo"
  >
    correct
  </div>
  <div
    className="sc-AxhCb JZaCe"
  >
    <ul
      className="sc-Axmtr dHjpTi"
    >
      <li
        className="sc-fzokOt gWWQxI"
      >
        <input
          checked={false}
          className="sc-fzqBZW cKxhfK"
          id="Kendrick Lamar"
          name="correct"
          onChange={[Function]}
          type="radio"
          value={0}
        />
        <label
          className="sc-fzqNJr VlIuW"
          htmlFor="Kendrick Lamar"
        >
          Kendrick Lamar
        </label>
      </li>
      <li
        className="sc-fzokOt gWWQxI"
      >
        <input
          checked={false}
          className="sc-fzqBZW cKxhfK"
          id="apple"
          name="correct"
          onChange={[Function]}
          type="radio"
          value={2}
        />
        <label
          className="sc-fzqNJr VlIuW"
          htmlFor="apple"
        >
          apple
        </label>
      </li>
    </ul>
  </div>
  <div
    className="sc-fzplWN buDbqD"
  >
    <button
      className="sc-AxmLO gmmVt button"
      disabled={true}
      onClick={[Function]}
    >
      skipped
    </button>
    <button
      className="sc-AxmLO sc-fzozJi jlLOkV button"
      disabled={true}
      onClick={[Function]}
    >
      discard
    </button>
  </div>
</div>
`;
